File: frontend/src/app/services/leave.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface LeaveRequest {
  id?: number;
  employeeId: number;
  employeeName?: string;
  department?: string;
  startDate: string;
  endDate: string;
  type: string;
  status?: string;
  reason: string;
  adminRemarks?: string;
  days?: number;
  createdAt?: string;
}

@Injectable({ providedIn: 'root' })
export class LeaveService {
  private apiUrl = 'http://localhost:8080/api/leaves';

  constructor(private http: HttpClient) {}

  applyForLeave(leave: LeaveRequest): Observable<LeaveRequest> {
    return this.http.post<LeaveRequest>(this.apiUrl, leave);
  }

  getLeavesByEmployee(employeeId: number): Observable<LeaveRequest[]> {
    return this.http.get<LeaveRequest[]>(`${this.apiUrl}/employee/${employeeId}`);
  }

  getAllLeaves(): Observable<LeaveRequest[]> {
    return this.http.get<LeaveRequest[]>(this.apiUrl);
  }

  getPendingLeaves(): Observable<LeaveRequest[]> {
    return this.http.get<LeaveRequest[]>(`${this.apiUrl}/pending`);
  }

  updateLeaveStatus(id: number, status: string, remarks?: string): Observable<LeaveRequest> {
    let url = `${this.apiUrl}/${id}/status?status=${status}`;
    if (remarks) {
      url += `&remarks=${encodeURIComponent(remarks)}`;
    }
    return this.http.patch<LeaveRequest>(url, {});
  }

  withdrawLeave(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getLeaveSummary(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/summary`);
  }
}

File: backend/src/main/java/com/workforce/workforcehub/service/LeaveService.java
package com.workforce.workforcehub.service;

import com.workforce.workforcehub.dto.LeaveApplicationRequest;
import com.workforce.workforcehub.dto.LeaveDto;
import com.workforce.workforcehub.entity.Employee;
import com.workforce.workforcehub.entity.LeaveRequest;
import com.workforce.workforcehub.repository.EmployeeRepository;
import com.workforce.workforcehub.repository.LeaveRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class LeaveService {
    
    private final LeaveRepository leaveRepository;
    private final EmployeeRepository employeeRepository;
    
    @Transactional
    public LeaveDto applyForLeave(LeaveApplicationRequest request) {
        Employee employee = employeeRepository.findById(request.getEmployeeId())
                .orElseThrow(() -> new RuntimeException("Employee not found"));
                
        if (request.getStartDate().isAfter(request.getEndDate())) {
            throw new RuntimeException("Start date cannot be after end date");
        }
        
        LeaveRequest leave = new LeaveRequest();
        leave.setEmployee(employee);
        leave.setStartDate(request.getStartDate());
        leave.setEndDate(request.getEndDate());
        leave.setType(request.getType());
        leave.setReason(request.getReason());
        leave.setStatus("PENDING");
        
        LeaveRequest saved = leaveRepository.save(leave);
        return mapToDto(saved);
    }
    
    public List<LeaveDto> getLeavesByEmployee(Long employeeId) {
        return leaveRepository.findByEmployeeIdOrderByCreatedAtDesc(employeeId)
                .stream().map(this::mapToDto).collect(Collectors.toList());
    }
    
    public List<LeaveDto> getAllLeaves() {
        return leaveRepository.findAllByOrderByCreatedAtDesc()
                .stream().map(this::mapToDto).collect(Collectors.toList());
    }
    
    public List<LeaveDto> getPendingLeaves() {
        return leaveRepository.findByStatusOrderByCreatedAtDesc("PENDING")
                .stream().map(this::mapToDto).collect(Collectors.toList());
    }
    
    @Transactional
    public LeaveDto updateLeaveStatus(Long id, String status, String remarks) {
        LeaveRequest leave = leaveRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Leave request not found"));
                
        leave.setStatus(status);
        if (remarks != null) {
            leave.setAdminRemarks(remarks);
        }
        
        return mapToDto(leaveRepository.save(leave));
    }
    
    public java.util.Map<String, Long> getLeaveSummary() {
        java.util.Map<String, Long> summary = new java.util.HashMap<>();
        List<LeaveRequest> allLeaves = leaveRepository.findAll();
        
        long pending = allLeaves.stream().filter(l -> "PENDING".equals(l.getStatus())).count();
        long approved = allLeaves.stream().filter(l -> "APPROVED".equals(l.getStatus())).count();
        long rejected = allLeaves.stream().filter(l -> "REJECTED".equals(l.getStatus())).count();
        
        summary.put("pending", pending);
        summary.put("approved", approved);
        summary.put("rejected", rejected);
        summary.put("total", (long) allLeaves.size());
        
        return summary;
    }
    
    private LeaveDto mapToDto(LeaveRequest leave) {
        LeaveDto dto = new LeaveDto();
        dto.setId(leave.getId());
        dto.setEmployeeId(leave.getEmployee().getId());
        dto.setEmployeeName(leave.getEmployee().getName());
        dto.setDepartment(leave.getEmployee().getDepartment());
        dto.setStartDate(leave.getStartDate());
        dto.setEndDate(leave.getEndDate());
        dto.setType(leave.getType());
        dto.setStatus(leave.getStatus());
        dto.setReason(leave.getReason());
        dto.setAdminRemarks(leave.getAdminRemarks());
        
        long days = ChronoUnit.DAYS.between(leave.getStartDate(), leave.getEndDate()) + 1;
        dto.setDays(days);
        dto.setCreatedAt(leave.getCreatedAt() != null ? leave.getCreatedAt().toString() : null);
        return dto;
    }

    @Transactional
    public void withdrawLeave(Long id) {
        LeaveRequest leave = leaveRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Leave request not found"));
        leaveRepository.delete(leave);
    }
}
